src/webgl-preview.ts

```diff
@@ -772,47 +398,64 @@ export class WebGLPreview {
-      evt.stopPropagation();
-      evt.preventDefault();
-      if (evt.dataTransfer) evt.dataTransfer.dropEffect = 'copy';
-      this.canvas.classList.add('dragging');
-    });
+  private renderLines(travels = this.job.travels(), extrusions = this.job.extrusions()): void {
+    if (this.renderTravel) {
+      const material = new LineBasicMaterial({ color: this._travelColor, linewidth: this.lineWidth });
+      this.disposables.push(material);
 
-    this.canvas.addEventListener('dragleave', (evt) => {
-      evt.stopPropagation();
-      evt.preventDefault();
-      this.canvas.classList.remove('dragging');
-    });
+      travels.forEach((path) => {
+        const geometry = path.line();
+        const line = new LineSegments(geometry, material);
+        this.group?.add(line);
+      });
+    }
 
-    this.canvas.addEventListener('drop', async (evt) => {
-      evt.stopPropagation();
-      evt.preventDefault();
-      this.canvas.classList.remove('dragging');
-      const files: FileList | [] = evt.dataTransfer?.files ?? [];
-      const file = files[0];
+    if (this.renderExtrusion && !this.renderTubes) {
+      const lineMaterials = {} as Record<number, LineBasicMaterial>;
 
-      this.clear();
+      if (Array.isArray(this._extrusionColor)) {
+        this._extrusionColor.forEach((color, index) => {
+          lineMaterials[index] = new LineBasicMaterial({ color, linewidth: this.lineWidth });
+        });
+      } else {
+        lineMaterials[0] = new LineBasicMaterial({
+          color: this._extrusionColor,
+          linewidth: this.lineWidth
+        });
+      }
 
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      await this._readFromStream(file.stream() as unknown as ReadableStream<any>);
-      this.render();
-    });
+      extrusions.forEach((path) => {
+        const geometry = path.line();
+        const line = new LineSegments(geometry, lineMaterials[path.tool]);
+        this.group?.add(line);
+      });
+    }
   }
 
-  private batchGeometries() {
-    if (this._geometries) {
-      for (const color in this._geometries) {
-        const batchedMesh = this.createBatchMesh(parseInt(color));
-        while (this._geometries[color].length > 0) {
-          const geometry = this._geometries[color].pop();
-          const geometryId = batchedMesh.addGeometry(geometry);
-          batchedMesh.addInstance(geometryId);
+  private renderGeometries(paths = this.job.extrusions()): void {
+    if (Object.keys(this._geometries).length === 0 && this.renderTubes) {
+      let color: number;
+      paths.forEach((path) => {
+        if (Array.isArray(this._extrusionColor)) {
+          color = this._extrusionColor[path.tool].getHex();
+        } else {
+          color = this._extrusionColor.getHex();
         }
-      }
+
+        this._geometries[color] ||= [];
+        this._geometries[color].push(path.geometry());
+      });
     }
+
+    for (const color in this._geometries) {
```

Comment:
brilliant!

Reply:
This is very similar to what we have in v2. Current users benefit from it!